tenant/storage_bucket.go

```diff
@@ -0,0 +1,418 @@
+		return nil, ErrBucketNotFound
+	}
+
+	if err != nil {
+		return nil, ErrInternalServiceError(err)
+	}
+
+	return unmarshalBucket(v)
+}
+
+func (s *Store) GetBucketByName(ctx context.Context, tx *Tx, orgID influxdb.ID, n string) (*influxdb.Bucket, error) {
+	key, err := bucketIndexKey(orgID, n)
+	if err != nil {
+		return nil, &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Err:  err,
+		}
+	}
+
+	idx, err := tx.Bucket(bucketIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := idx.Get(key)
+
+	// allow for hard coded bucket names that dont exist in the system
+	if kv.IsNotFound(err) {
+		switch n {
+		case influxdb.TasksSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.TasksSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.TasksSystemBucketName,
+				RetentionPeriod: influxdb.TasksSystemBucketRetention,
+				Description:     "System bucket for task logs",
+				OrgID:           orgID,
+			}, nil
+		case influxdb.MonitoringSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.MonitoringSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.MonitoringSystemBucketName,
+				RetentionPeriod: influxdb.MonitoringSystemBucketRetention,
+				Description:     "System bucket for monitoring logs",
+				OrgID:           orgID,
+			}, nil
+		default:
+			return nil, ErrBucketNotFound
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var id influxdb.ID
+	if err := id.Decode(buf); err != nil {
+		return nil, &influxdb.Error{
+			Err: err,
+		}
+	}
+	return s.GetBucket(ctx, tx, id)
+}
+
+type BucketFilter struct {
+	Name           *string
+	OrganizationID *influxdb.ID
+}
+
+func (s *Store) ListBuckets(ctx context.Context, tx *Tx, filter BucketFilter, opt ...influxdb.FindOptions) ([]*influxdb.Bucket, error) {
+	// this isn't a list action its a `GetBucketByName`
+	if (filter.OrganizationID != nil && filter.OrganizationID.Valid()) && filter.Name != nil {
+		return nil, invalidBucketListRequest
+	}
+
+	// if we dont have any options it would be irresponsible to just give back all orgs in the system
+	if len(opt) == 0 {
+		opt = append(opt, influxdb.FindOptions{
+			Limit: influxdb.DefaultPageSize,
+		})
+	}
+	o := opt[0]
+	if o.Limit > influxdb.MaxPageSize || o.Limit == 0 {
+		o.Limit = influxdb.MaxPageSize
+	}
+
+	// if an organization is passed we need to use the index
+	if filter.OrganizationID != nil {
+		return s.listBucketsByOrg(ctx, tx, *filter.OrganizationID, o)
+	}
+
+	b, err := tx.Bucket(bucketBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	cursor, err := b.ForwardCursor(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer cursor.Close()
+
+	count := 0
+	bs := []*influxdb.Bucket{}
+	for k, v := cursor.Next(); k != nil; k, v = cursor.Next() {
+		if o.Offset != 0 && count < o.Offset {
+			count++
+			continue
+		}
+		b, err := unmarshalBucket(v)
+		if err != nil {
+			continue
```

Comment:
Out of interest why don't we care about this error? if we do not, do we want to log or comment as to why?